backend/src/main/java/com/melodify/melodify/service/PostService.java

```diff
@@ -0,0 +1,110 @@
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+public class PostService {
+
+    private final PostRepository postRepository;
+    private final UserService userService;
+    private final TagService tagService;
+
+    public PostService(PostRepository postRepository, UserService userService, TagService tagService) {
+        this.postRepository = postRepository;
+        this.userService = userService;
+        this.tagService = tagService;
+    }
+
+    private List<PostResponse> mapper(List<Post> posts) {
+        return posts.stream().map(post -> {
+            List<String> tags = tagService.getTagsByPost(post).stream().map(Tag::getTag).collect(Collectors.toList());
+            return new PostResponse(post, tags);
+        }).collect(Collectors.toList());
+    }
+
+    public List<PostResponse> getAllPosts() {
+        return mapper(postRepository.findAll());
+    }
+
+    public List<PostResponse> getPostsByAuthorAndTag(String author, String tag) {
+        List<PostResponse> temp = getPostsByAuthor(author);
+        return temp.stream().filter(post -> post.getTags().contains(tag)).collect(Collectors.toList());
+    }
+
+    public List<PostResponse> getPostsByAuthor(String author) {
+        User user = userService.getOneUserByUsername(author);
+        if (user == null) {
+            throw new IllegalArgumentException("User not found");
+        }
+        return mapper(postRepository.findByAuthor(user));
+    }
+
+    public List<PostResponse> getPostsByTag(String tag) {
+        return mapper(tagService.getPostsByTag(tag));
+    }
+
+    public long createOnePost(PostCreateRequest newPostRequest) {
+        User author = userService.getOneUserByUsername(newPostRequest.getAuthor());
+        if (author == null) {
+            throw new IllegalArgumentException("Author not found");
+        }
+        Post post = new Post(newPostRequest.getText(), author, newPostRequest.getMedia_url());
+        long id = postRepository.save(post).getId();
+        try {
+            for (String tag : newPostRequest.getTags()) {
+                tagService.createOneTag(tag, post);
+            }
+        } catch (Exception e) {
+            postRepository.delete(post);
+            throw new IllegalArgumentException(e.getMessage());
+        }
+        return id;
+    }
+
+    private PostResponse convertToResponse(Post post) {
+        List<String> tags = tagService.getTagsByPost(post).stream().map(Tag::getTag).collect(Collectors.toList());
+        return new PostResponse(post, tags);
+    }
+
+    public PostResponse getOnePostById(Long postId) {
+        Post post = postRepository.findById(postId)
+                .orElseThrow(() -> new RuntimeException("Post not found with ID: " + postId));
+        return convertToResponse(post);
+    }
+
+    public PostResponse updateOnePostById(Long postId, PostUpdateRequest updatePost) {
+        Post post = postRepository.findById(postId).orElse(null);
```

Comment:
Check if the author of the post is the same who sent this request, with authorization token.